src/main/java/Duke.java

```diff
@@ -1,10 +1,57 @@
+import java.nio.file.Paths;
+import java.util.Scanner;
+
 public class Duke {
+    private static final String DATABASE_DIRECTORY_PATH = Paths.get(System.getProperty("user.dir"),
+            "../../../", "data").toString();
+    private Parser parser;
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
+
+    /**
+     * */
+    public Duke(String databaseDirectoryPath) {
+        this.ui = new Ui();
+        this.storage = new Storage(databaseDirectoryPath);
+        this.parser = new Parser();
+
+        try {
+            this.tasks = new TaskList(storage.load());
+        } catch (DukeException e) {
+            this.ui.showErrorMessage("Could not load saved tasks");
+            this.tasks = new TaskList();
+        }
+    }
+
+    /**
+     * Starts the bot.
+     */
+    public void run() {
+        this.ui.greetUser();
+
+        boolean keepApplicationRunning = true;
+        Scanner sc = new Scanner(System.in);
+
+        while (keepApplicationRunning) {
+            try {
+                String userInput = this.ui.getUserInput(sc);
+                if (userInput.equals("bye")) {
+                    keepApplicationRunning = false;
+                }
+                this.parser.parseCommands(this.tasks, userInput, sc);
```

Comment:
"bye" could technically be treated as a command. Having the "bye" command case here instead of in line 42 might cause both an inconsistency in abstraction levels and deep nesting. Perhaps you could change the `Parser.parseCommands()` method to also factor in the "bye" case, in which it would call `sc.close()`? 
This would also set `sc.hasNext()` to `false`, which naturally breaks the loop and ties in with the earlier comment.

Reply:
This is a nice solution! It certainly adheres more to the abstraction levels than the current code! Thanks for pointing it out!